### ISSUE TITLE
Image Metadata/Properties Type

### ISSUE THREAD

**Issue by jtopjian** [gophercloud/gophercloud#216]
I'm still researching this one, but it looks as though image metadata and properties can hold more than `map[string]string`.

This was originally discovered [here](https://github.com/hashicorp/terraform/issues/10767) and per the output [here](https://github.com/hashicorp/terraform/issues/10767#issuecomment-267507276),  the properties of the image contains a nested array.

I was able to reproduce the reported issue in question by setting similar metadata/properties on an existing image:

```shell
$ openstack --debug image set --property block_device_mapping='[{"device_name": "/dev/vdb", "boot_index": -1, "source_type": "blank", "destination_type": "local", "volume_size": 1, "delete_on_termination": true}]' jttest
```

Similarly, I was able to add arbitrary arrays and hashes/maps as properties:

```shell
$ openstack --debug image set --property foo='[{"a": "b"}]' jttest
$ openstack --debug image set --property bar='["a","b","c"]' jttest
$ openstack --debug image set --property baz='{"a": "b"}' jttest
```

So it doesn't seem like the API is enforcing a schema for properties like this (meaning: `block_device_mapping` isn't an exception to the otherwise `map[string]string`).

When Gophercloud hits a property like this, the following error is triggered:

```
Unexpected error from Get: json: cannot unmarshal array into Go value of type string
```

This is reproducible both in unit and acceptance tests.

Changing `map[string]string` to `map[string]interface{}` quickly resolves the issue, but I'm unsure if that's the best solution or if that's _all_ that is required to fix this.

Regardless of the fix, I believe both the `compute/v2/images` and `images/v2/images` packages will need updated.

TODO (for either myself or if anyone else wants to do this):

* Dig into the Glance code and find where it's setting metadata/properties
* Determine the difference between "metadata" and "properties". The `images/v2/images` defines both, but the `openstack` command-line client only allows updates to `properties`. `properties` are displayed as `metadata` through the Compute API (which I understand will be deprecated soon).

**Comment by dvianello** [comment]
Hi @jtopjian, any news on this? We're currently hit pretty badly by this bug at the moment!

**Comment by jtopjian** [comment]
I'll admit to total procrastination on this one. I'll make some time for this soon. :)

**Comment by jtopjian** [comment]
Good news: This is, in fact, a special case for `block_device_mapping` and only in Compute Image API:

https://github.com/openstack/nova/blob/master/nova/image/glance.py#L672-L721

Bad news: I'm not sure how to fix it. Going to post my work in progress.

**Comment by jrperritt** [comment]
closed by #285
